### PR TITLE
[3216][ADD] website_product_visibility_control_timer

### PR DESCRIPTION
**Pull request by AungKoKoLin1997**
[3216](https://www.quartile.co/web#id=3216&action=771&model=project.task&view_type=form&menu_id=505)

### CODE REVIEW

File: website_product_visibility_control_timer/models/product_template.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+from odoo import fields, models
+
+
+class ProductTemplate(models.Model):
+    _inherit = "product.template"
+
+    expected_publish_date = fields.Datetime()
+    expected_unpublish_date = fields.Datetime()
+
+    def _check_website_product_visibility(self):
```
Comment (yostashiro) [inline website_product_visibility_control_timer/models/product_template.py]: Add methods to return publish domain and unpublish domain for extensibility.

File: website_product_visibility_control_timer/readme/DESCRIPTION.rst [hunk 2]
```diff
@@ -0,0 +1,3 @@
+This module does the following:
+
+- Control the visibility of products in website depend on their publish date and unpublish date.
```
Comment (yostashiro) [inline website_product_visibility_control_timer/readme/DESCRIPTION.rst]:
```suggestion
This module adds expected publish/unpublish date fields to the product, and a scheduled action to publish/unpublish products according to the product settings.

This function can be useful if you run an eCommerce platform where a set of products need to be sold for a scheduled timeframe.
```

File: website_product_visibility_control_timer/models/product_template.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        publish_products = self.env["product.template"].search(
+            [
+                ("website_published", "=", False),
+                ("expected_publish_date", "<=", fields.datetime.now()),
+                "|",
+                ("expected_unpublish_date", ">", fields.datetime.now()),
+                ("expected_unpublish_date", "=", False),
+            ]
```
Comment (yostashiro) [inline website_product_visibility_control_timer/models/product_template.py]: This domain does not work if user intends to first unpublish an already published product, and then unpublish it.  I know this scenario sounds rather unlikely, but can we handle this case also?

File: website_product_visibility_control_timer/models/product_template.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        )
+        publish_products.write({"website_published": True})
+
+        unpublish_products = self.env["product.template"].search(
+            [
+                ("website_published", "=", True),
+                ("expected_unpublish_date", "<=", fields.datetime.now()),
+            ]
```
Comment (yostashiro) [inline website_product_visibility_control_timer/models/product_template.py]: Same as above.

**Review by yostashiro — CHANGES_REQUESTED** [review]
Can we change the module name to website_product_publish_cron, and propose it in OCA?

**Comment by AungKoKoLin1997** [comment]
This module was proposed to OCA and the next activity will be followed in OCA PR.

**Pull request closed without merging**